### ISSUE TITLE
torch.aten.native_layer_norm to linalg

### ISSUE THREAD

**Issue by saienduri** [nod-ai/SHARK-Turbine#569]
We are having following problems with onnx lowering into tensor.expand_shape in these models (beit-base-patch16-224-pt22k-ft22k, deit-small-distilled-patch16-224, vit-base-patch16-224). Repro instructions can be found [here](https://github.com/nod-ai/SHARK-Turbine/issues/564)::

```
beit-base-patch16-224-pt22k-ft22k.default.pytorch.torch.mlir:336:36: error: 'tensor.expand_shape' op invalid to have a single dimension (0) expanded into multiple dynamic dims (0,1)
    %result0, %result1, %result2 = torch.aten.native_layer_norm %293, %294, %10, %11, %float9.999990e-13 : !torch.vtensor<[?,?,768],f32>, !torch.list<int>, !torch.vtensor<[768],f32>, !torch.vtensor<[768],f32>, !torch.float -> !torch.vtensor<[?,?,768],f32>, !torch.vtensor<[?,?,1],f32>, !torch.vtensor<[?,?,1],f32>
                                   ^
beit-base-patch16-224-pt22k-ft22k.default.pytorch.torch.mlir:336:36: note: see current operation: %295 = "tensor.expand_shape"(%294) <{reassociation = [[0, 1]]}> : (tensor<?xf32>) -> tensor<?x?xf32>
```

```
deit-small-distilled-patch16-224.default.pytorch.torch.mlir:352:36: error: 'tensor.expand_shape' op expected dimension 0 of collapsed type to be dynamic since one or more of the corresponding dimensions in the expanded type is dynamic
    %result0, %result1, %result2 = torch.aten.native_layer_norm %297, %298, %11, %12, %float9.999990e-13 : !torch.vtensor<[?,198,384],f32>, !torch.list<int>, !torch.vtensor<[384],f32>, !torch.vtensor<[384],f32>, !torch.float -> !torch.vtensor<[?,198,384],f32>, !torch.vtensor<[?,198,1],f32>, !torch.vtensor<[?,198,1],f32>
                                   ^
deit-small-distilled-patch16-224.default.pytorch.torch.mlir:352:36: note: see current operation: %263 = "tensor.expand_shape"(%262) <{reassociation = [[0, 1]]}> : (tensor<198xf32>) -> tensor<?x198xf32>
```

```
vit-base-patch16-224.default.pytorch.torch.mlir:316:36: error: 'tensor.expand_shape' op expected dimension 0 of collapsed type to be dynamic since one or more of the corresponding dimensions in the expanded type is dynamic
    %result0, %result1, %result2 = torch.aten.native_layer_norm %272, %273, %10, %11, %float9.999990e-13 : !torch.vtensor<[?,197,768],f32>, !torch.list<int>, !torch.vtensor<[768],f32>, !torch.vtensor<[768],f32>, !torch.float -> !torch.vtensor<[?,197,768],f32>, !torch.vtensor<[?,197,1],f32>, !torch.vtensor<[?,197,1],f32>
                                   ^
vit-base-patch16-224.default.pytorch.torch.mlir:316:36: note: see current operation: %259 = "tensor.expand_shape"(%258) <{reassociation = [[0, 1]]}> : (tensor<197xf32>) -> tensor<?x197xf32>
```

**Comment by renxida** [comment]
Reproing with

HF_TOKEN=... python run.py --torchmlirbuild /path/to/torch-mlir/build --ireebuild /path/to/iree-build --cachedir ~/.cache/huggingface --tests pytorch/models/deit-small-distilled-patch16-224 -r test-onnx --tolerance .001 .001 --mode onnx --report

**Comment by renxida** [comment]
Minimal repro example:

```
func.func @native_layer_norm(%input: !torch.vtensor<[?,198,384],f32>, %weight: !torch.vtensor<[384],f32>, %bias: !torch.vtensor<[384],f32>, %eps: !torch.float) -> (!torch.vtensor<[?,198,384],f32>, !torch.vtensor<[?,198,1],f32>, !torch.vtensor<[?,198,1],f32>) {
  %int384 = torch.constant.int 384
  %normalized_shape = torch.prim.ListConstruct %int384 : (!torch.int) -> !torch.list<int>
  %result0, %result1, %result2 = torch.aten.native_layer_norm %input, %normalized_shape, %weight, %bias, %eps : !torch.vtensor<[?,198,384],f32>, !torch.list<int>, !torch.vtensor<[384],f32>, !torch.vtensor<[384],f32>, !torch.float -> !torch.vtensor<[?,198,384],f32>, !torch.vtensor<[?,198,1],f32>, !torch.vtensor<[?,198,1],f32>
  return %result0, %result1, %result2 : !torch.vtensor<[?,198,384],f32>, !torch.vtensor<[?,198,1],f32>, !torch.vtensor<[?,198,1],f32>
}
```

For some reason no compile error on this. 
### I tried the following:

**Succeedes without an issue:**
`/home/azureuser/iree-build/tools/iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu  minimal.mlir > minimal.vmfb`

**Comment by renxida** [comment]
Repro log with:
- original
- stripped
- minimal example

```
+ echo --- reproing with original model mlir ---
--- reproing with original model mlir ---
+ /home/azureuser/iree-build/tools/iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu deit-small-distilled-patch16-224.default.pytorch.torch.mlir -o deit-small-distilled-patch16-224.default.vmfb
/home/azureuser/iree-build/tools/iree-compile: /home/azureuser/miniconda/lib/libtinfo.so.6: no version information available (required by /home/azureuser/iree-build/lib/libIREECompiler.so)
deit-small-distilled-patch16-224.default.pytorch.torch.mlir:364:36: error: 'tensor.expand_shape' op expected dimension 0 of collapsed type to be dynamic since one or more of the corresponding dimensions in the expanded type is dynamic
    %result0, %result1, %result2 = torch.aten.native_layer_norm %309, %310, %11, %12, %float9.999990e-13 : !torch.vtensor<[?,198,384],f32>, !torch.list<int>, !torch.vtensor<[384],f32>, !torch.vtensor<[384],f32>, !torch.float -> !torch.vtensor<[?,198,384],f32>, !torch.vtensor<[?,198,1],f32>, !torch.vtensor<[?,198,1],f32>
                                   ^
deit-small-distilled-patch16-224.default.pytorch.torch.mlir:364:36: note: see current operation: %263 = "tensor.expand_shape"(%262) <{reassociation = [[0, 1]]}> : (tensor<198xf32>) -> tensor<?x198xf32>
+ echo Return code: 1
Return code: 1
+ echo --- reproing with stripped model mlir ---
--- reproing with stripped model mlir ---
+ /home/azureuser/iree-build/tools/iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu stripped/deit-small-distilled-patch16-224.default.pytorch.torch.stripped.mlir -o deit-small-distilled-patch16-224.default.stripped.vmfb
/home/azureuser/iree-build/tools/iree-compile: /home/azureuser/miniconda/lib/libtinfo.so.6: no version information available (required by /home/azureuser/iree-build/lib/libIREECompiler.so)
stripped/deit-small-distilled-patch16-224.default.pytorch.torch.stripped.mlir:364:36: error: 'tensor.expand_shape' op expected dimension 0 of collapsed type to be dynamic since one or more of the corresponding dimensions in the expanded type is dynamic
    %result0, %result1, %result2 = torch.aten.native_layer_norm %309, %310, %11, %12, %float9.999990e-13 : !torch.vtensor<[?,198,384],f32>, !torch.list<int>, !torch.vtensor<[384],f32>, !torch.vtensor<[384],f32>, !torch.float -> !torch.vtensor<[?,198,384],f32>, !torch.vtensor<[?,198,1],f32>, !torch.vtensor<[?,198,1],f32>
                                   ^
stripped/deit-small-distilled-patch16-224.default.pytorch.torch.stripped.mlir:364:36: note: see current operation: %73 = "tensor.expand_shape"(%72) <{reassociation = [[0, 1]]}> : (tensor<198xf32>) -> tensor<?x198xf32>
+ echo Return code: 1
Return code: 1
+ echo --- reproing with minimal example ---
--- reproing with minimal example ---
+ /home/azureuser/iree-build/tools/iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu minimal.mlir -o minimal.vmfb
/home/azureuser/iree-build/tools/iree-compile: /home/azureuser/miniconda/lib/libtinfo.so.6: no version information available (required by /home/azureuser/iree-build/lib/libIREECompiler.so)
error opening input file: cannot open input file 'minimal.mlir': No such file or directory
+ echo Return code: 1
Return code: 1
```

**Comment by renxida** [comment]
IR dump after failure (with failing line highlighted)
https://gist.github.com/renxida/fdcec365e89ae317fdd8545fbd4ddaf8#file-gistfile0-txt-L117


Relevant section:
```
  %72 = "linalg.generic"(%69, %71) <{indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = [#linalg.iterator_type<parallel>, #linalg.iterator_type<reduction>], operandSegmentSizes = array<i32: 1, 1>}> ({
  ^bb0(%arg3: f32, %arg4: f32):
    %1595 = "arith.addf"(%arg3, %arg4) <{fastmath = #arith.fastmath<none>}> : (f32, f32) -> f32
    "linalg.yield"(%1595) : (f32) -> ()
  }) : (tensor<198x384xf32>, tensor<198xf32>) -> tensor<198xf32>
  %73 = "tensor.expand_shape"(%72) <{reassociation = [[0, 1]]}> : (tensor<198xf32>) -> tensor<?x198xf32>
  %74 = "tensor.empty"(%53) : (index) -> tensor<?x198xf32>
  %75 = "linalg.generic"(%73, %74) <{indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = [#linalg.iterator_type<parallel>, #linalg.iterator_type<parallel>], operandSegmentSizes = array<i32: 1, 1>}> ({
```

**Comment by renxida** [comment]
TLDR: Looks like we generate the problematic tensor.expand_shape during a `FoldUnitExtentDims` pass

Print-ir-after-all produces a 15MB text file.
```
/home/azureuser/iree-build/tools/iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu  stripped/deit-small-distilled-patch16-224.default.pytorch.torch.stripped.mlir -o deit-small-distilled-patch16-224.default.stripped.vmfb   --mlir-pretty-debuginfo         --mlir-print-ir-after-all &> printirafterall.mlir
```

But we can grep
```
┌─[0]─[azureuser@xida-cpu-0]─[~/SHARK-TestSuite/e2eshark/test-onnx/pytorch/models/deit-small-distilled-patch16-224]
└──╼ $cat printirafterall.mlir | grep -E '\(tensor<198xf32>\)|IR Dump'

<... a bunch of ir dump after messages with no matches to the tensor<198xf32> pattern ...>

// -----// IR Dump After ConvertElementwiseToLinalgPass (convert-elementwise-to-linalg) //----- //
// -----// IR Dump After RaiseSpecialOps (iree-global-opt-raise-special-ops) //----- //
// -----// IR Dump After DecomposeConcat (iree-global-opt-decompose-concat) //----- //
// -----// IR Dump After GeneralizeLinalgNamedOps (iree-global-opt-generalize-linalg-named-ops) //----- //
// -----// IR Dump After FoldUnitExtentDims Failed (iree-flow-fold-unit-extent-dims) //----- //
  %73 = "tensor.expand_shape"(%72) <{reassociation = [[0, 1]]}> : (tensor<198xf32>) -> tensor<?x198xf32>
  %86 = "tensor.expand_shape"(%85) <{reassociation = [[0, 1]]}> : (tensor<198xf32>) -> tensor<?x198xf32>
stripped/deit-small-distilled-patch16-224.default.pytorch.torch.stripped.mlir:364:36: note: see current operation: %73 = "tensor.expand_shape"(%72) <{reassociation = [[0, 1]]}> : (tensor<198xf32>) -> tensor<?x198xf32>
```

**Comment by renxida** [comment]
With the minimal ir, it correctly preserves the <?x198xf32>. See output at: https://gist.github.com/renxida/e4347a1ef027e9bf7ed3487e7d87d577


Command:
```
/home/azureuser/iree-build/tools/iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu  stripped/minimal.mlir -o minimal.vmfb   --mlir-pretty-debuginfo --mlir-print-ir-before=iree-flow-fold-unit-extent-dims --mlir-print-ir-after=iree-flow-fold-unit-extent-dims |& gh gist create - -d "minimal ir before and after FoldUnitExtentDims"
```

**Comment by renxida** [comment]
A search of the error message indicates that the error is generated by

https://github.com/shark-infra/llvm-project/blob/b3291793f11924a3b62601aabebebdcfbb12a9a1/mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp#L250-L257

This function `LogicalResult mlir::reshapeLikeShapesAreCompatible` is used for both the tensor.expand_shape and tensor.collapse_shape ops.

But this particular check 
```cpp
    if (dynamicShape) {
      if (!ShapedType::isDynamic(collapsedShape[map.index()])) {
        return emitError(
            "expected dimension " + Twine(map.index()) +
            " of collapsed type to be dynamic since one or more of the "
            "corresponding dimensions in the expanded type is dynamic");
      }
    }
```

may not make sense - - i can see a world where we allow collapsing unknown-shape dims.

**Comment by renxida** [comment]
https://gist.github.com/renxida/d39d9dd77991836d620737e57161d4a3

--compile-to=input

**Comment by renxida** [comment]
previously i thought the problem is introduced by `iree-flow-fold-unit-extent-dims` because the `expand_shape` wasn't there before the pass and after the pass it was there.

but then after slogging through it a bunch with stanley i noticed a couple of weird things in the ir from right before expand_shape:

1. a linalg generic that performed a reduce-sum over the last dim of a tensor, yet converted dim 0 from 1 to ?: `tensor<1x198x384xf32>) outs(%27 : tensor<?x198x1xf32>` (at: https://gist.github.com/renxida/cff78af3cc3dfde1e7755a8a6cdf24b9#file-before-expand-shape-mlir-L113) [^1]


2. a cast that turned a tensor with known shapes to unknown shapes `%cast = tensor.cast %8 : tensor<1x196x384xf32> to tensor<?x?x?xf32>` (see: https://gist.github.com/renxida/cff78af3cc3dfde1e7755a8a6cdf24b9#file-before-expand-shape-mlir-L64)



[^1]: full linalg generic here
```
  %28 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, 0)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%25 : tensor<1x198x384xf32>) outs(%27 : tensor<?x198x1xf32>) {
  ^bb0(%in: f32, %out: f32):
      %744 = arith.addf %in, %out : f32 // doing some sort of accumulative sum. adding output of previous loop. reduction. 
      linalg.yield %744 : f32
  } -> tensor<?x198x1xf32>
```

**Comment by renxida** [comment]
looks like (1) from the last comment is introduced after `iree-global-opt-decompose-concat`

https://gist.github.com/renxida/5eeeaa252692438c560b000f6e58421d

**Comment by renxida** [comment]
Linalg generics that turn known shapes into unknown shapes:

```
cat printirafterall.mlir | grep -E 'tensor<1x198x?x.*outs.*\?x198|IR Dump'
```

Result:
```mlir
// -----// IR Dump After ConvertTorchToLinalg (convert-torch-to-linalg) //----- //
  %533 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%cast_641, %cast_153 : tensor<1x198x384xf32>, tensor<?x198x384xf32>) outs(%532 : tensor<?x198x384xf32>) {
  %855 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%cast_1398, %cast_906 : tensor<1x198x384xf32>, tensor<?x198x384xf32>) outs(%854 : tensor<?x198x384xf32>) {
```

there are things like this that come relatively early in the pipeline (right after convert-torch-to-linalg).

These make sense because they're broadcasting an elementwise op between a 1x198x384 and a ?x194x384 to produce a ?x198x384

What doesn't make sense is when we have a `1x198x384` converted to a `?x198x1` with a single-operand reduce sum.

Need to filter specifically for those.

cat printirafterall.mlir | grep -E 'ins\(%\d+tensor<1x198x?x.*outs.*\?x198|IR Dump'

**Comment by renxida** [comment]
![image](https://github.com/nod-ai/SHARK-Turbine/assets/10362952/4b99dd44-5205-4fb7-88e5-18048964a6eb)

**Comment by renxida** [comment]
Let C2 denote the canonicalization pass after ExpandOps above.

C2 produces something likes this that eventually causes the error:

```
  %100 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, 0)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%97 : tensor<1x198x384xf32>) outs(%99 : tensor<?x198x1xf32>) {
  ^bb0(%in: f32 loc("after-memref-expand.mlir":587:8), %out: f32 loc("after-memref-expand.mlir":587:18)):
    %1984 = arith.addf %in, %out : f32 loc("after-memref-expand.mlir":588:13)
    linalg.yield %1984 : f32 loc("after-memref-expand.mlir":589:5)
  } -> tensor<?x198x1xf32> loc("after-memref-expand.mlir":586:10)
```

which, before canonicalization, looked like this:

```  %338 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, 0)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%cast_173 : tensor<?x198x384xf32>) outs(%337 : tensor<?x?x1xf32>) {
  ^bb0(%in: f32, %out: f32):
    %4260 = arith.addf %in, %out : f32
    linalg.yield %4260 : f32
  } -> tensor<?x?x1xf32>
```

which means this is an issue caused by an incomplete shape inference that only assigned 1 to dim 0 of the input but not the output.

**Comment by renxida** [comment]
# Summary so far:

1. original error: lowering native_layer_norm generated a tensor.expand_shape expanding a tensor<198xf32> to tensor<?x198xf32> after the pass that eliminates unit-extent dims.
2. this tensor.expand_shape causes an error because expand_shape expects only ? to correspond to ? dims, and not expanding a dim out to a ?
3. the tensor.expand_shape is materialized in between when the following pair of linalg ops are lowered:
```
// reduce sum over the last dim
// iterator types = parallel, parallel, reduction
%28 = linalg.generic ... ins(%25 : tensor<1x198x384xf32>) outs(%27 : tensor<?x198x1xf32>) ...

// elementwise division by a scalar
// iterator types = all parallel
%29 = linalg.generic ... ins(%28 : tensor<?x198x1xf32>) outs(%26 : tensor<?x198x1xf32>) ... 
```
5. The root cause seems to be a linalg.generic that performs a reduce-sum with: `ins(%25 : tensor<1x198x384xf32>) outs(%27 : tensor<?x198x1xf32>)`. Note that the reduce sum is performed on the 384 extent dim, but the first dim has input 1 but output ?
6. This doesn't make sense because we don't really touch the 1st dim and if the input is a 1, the output would also have to be a 1.
7. The reduce sum originally looked like this:
`%338 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, 0)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%cast_173 : tensor<?x198x384xf32>) outs(%337 : tensor<?x?x1xf32>)`
and the ? for the input is filled in by shape inference, but the ? for the out remained a ?.

**Comment by renxida** [comment]
Latest status: after updating to the latest torch-mlir and iree, encountering an Unsqueeze error. Either something fixed this, or an Unsqueeze error masked this.

**Comment by AmosLewis** [comment]
https://github.com/nod-ai/e2eshark-reports/blob/main/2024-06-12/onnx_reports/statusreport.md
model pass based on this report
